### PR TITLE
Smoke tests for assorted plugins

### PR DESCRIPTION
**Pull request by ssbarnea**
This aims to create a tox environment that tests that installation of
popular plugins does not break pytest.

Related to https://github.com/pytest-dev/pytest-cov/issues/430

### CODE REVIEW

**Comment by nicoddemus** [comment]
@ssbarnea I've update the PR to include simple integration tests for `pytest-cov`, `pytest-django`, `pytest-html`, and `pytest-sugar`. I also fixed the issue you identified originally in `pytest-cov`.

We should backport the commit that adds the integration tests back to `6.0.x` after this gets merged.

Fix pytest-dev/pytest-cov#430.

**Comment by nicoddemus** [comment]
I think this is a good starting point, but if others feel some plugin is missing, please speak up!

File: tox.ini
Comment (graingert) [inline tox.ini:128]: thoughts on a `pytest-anyio` test?

Comment (nicoddemus) [reply tox.ini:128]: Sounds good, feel free to push a new test to this branch. 👍

**Comment by agronholm** [comment]
You need to ensure that the tests are actually run. Add a failing case with strict `xfail=True` to make sure. I learned this the hard way when I found out AnyIO's hypothesis support had been silently broken the whole time.

File: src/_pytest/python.py
Comment (graingert) [inline src/_pytest/python.py:171]: I ordered these alphabetically - maybe by github starts or pypi downloads would be better?

**Comment by ssbarnea** [comment]
You may want to look at the PYTEST_REQPASS feature added by pytest-plus which assures that an exact number of tests passed. That was aimed for CI/CD usage for avoiding accidents where tests were skipped.

File: tox.ini
Comment (graingert) [inline tox.ini:140]: I think we should test all the async frameworks simultaneously - they have all agreed to use marks to allow them to all interoperate in a single test run, so we should test that here

File: tox.ini
Comment (graingert) [inline tox.ini:129]: @altendky pytest-twisted ?

File: tox.ini
Comment (agronholm) [reply tox.ini:140]: At least as long as you don't use async fixtures. `pytest-asyncio` is known to appropriate those regardless of marks (see https://github.com/pytest-dev/pytest-asyncio/issues/124 for the issue).

File: src/_pytest/python.py
Comment (hugovk) [reply src/_pytest/python.py:171]: Alphabetic is best for when looking at it later, for example, to check if a new one is already present or not.

Also no need to look up stars or downloads from elsewhere.

**Comment by graingert** [comment]
@ssbarnea what's the scope of this you're happy to accept? I think as plugin maintainers become aware of this PR more will want included.

I don't want to see a situation where this scope creeps all the way to testing all of pypi before it's merged! So I think we should merge this sooner and rather than later so maintainers can add their own smoke tests as PRs directly to pytest, rather than this branch?

File: tox.ini
Comment (graingert) [reply tox.ini:140]: > At least as long as you don't use async fixtures. 

That's a very good point, we currently don't test any async fixtures here.

**Comment by graingert** [comment]
> installation of popular plugins does not break pytest.

how popular is popular? Can we nail down a criteria for inclusion?

File: src/_pytest/python.py
Comment (altendky) [reply src/_pytest/python.py:171]: Alphabetical seems good, even if it leaves 'me' at the end. `:]`

**Comment by altendky** [comment]
> You may want to look at the PYTEST_REQPASS feature added by pytest-plus which assures that an exact number of tests passed. That was aimed for CI/CD usage for avoiding accidents where tests were skipped.

Is that better than just requiring coverage in your tests?

File: tox.ini
Comment (altendky) [reply tox.ini:129]: May as well alpha-sort here too?  Also `pytest-qt`.

**Review by altendky — COMMENTED** [review]
The stated purpose is to test that pytest doesn't break.  Shouldn't the full pytest suite be run instead of just a trivial case for each plugin?

**Comment by ssbarnea** [comment]
xdist and cov are ok dor start. good point about avoiding scope creep. we can always add more later.

File: tox.ini
Comment (graingert) [inline tox.ini:132]:
```suggestion
commands =
    [ $(pip list -o | wc -l) -eq 2 ]
```

I think we want something like this to make sure there's no version constraints holding us back from the latest releases

/cc @McSinyx

Comment (McSinyx) [reply tox.ini:132]: As I mentioned [when I suggested this on IRC](http://kafka.dcpython.org/day/pypa/2020-09-13#15.06.54.McSinyx[m]), it's quite of a hack.  I'm not sure how stable pip's output will be but I think at least adding `--quiet` would make it safer.

File: tox.ini
Comment (graingert) [inline tox.ini:118]: how about?

```suggestion
[testenv:plugins]
pip_pre=true
```
https://tox.readthedocs.io/en/latest/config.html#conf-pip_pre

**Comment by nicoddemus** [comment]
> The stated purpose is to test that pytest doesn't break. Shouldn't the full pytest suite be run instead of just a trivial case for each plugin?

This makes sense, but I would rather start things slowly: 

1. Plugins might interact with pytest differently in a way that breaks the pytest suite, but are not wrong: for example `pytest-sugar` changes output, and several tests check output and would break, requiring us to special case tests, or choosing a subset.
2. We can always improve things later the more we get the feeling how this will work. Having it being a quick sanity check run is beneficial, and reliable.

> I don't want to see a situation where this scope creeps all the way to testing all of pypi before it's merged! So I think we should merge this sooner and rather than later so maintainers can add their own smoke tests as PRs directly to pytest, rather than this branch?

Agreed! This is already a large improvement over what we have today, so let's get this in as soon as we feel this is good enough.

File: tox.ini
Comment (ssbarnea) [reply tox.ini:132]: Just run `pip check` and it will fail if any conflicts are detected.

File: tox.ini
Comment (nicoddemus) [reply tox.ini:118]: To test pre-release of plugins you mean? I don't see many plugins doing that (never seen any TBH). You think it is worth it?

**Comment by nicoddemus** [comment]
Add 3 more: `pytest-flakes`, `pytest-qt` and `pytest-bdd`. Also added `pip check` as suggested.

**Comment by nicoddemus** [comment]
Hmm as I feared: pytest-qt, besides depending on a large library, also requires an X server. While it is possible, I don't want to introduce complicated setups at this stage.

File: tox.ini
Comment (altendky) [reply tox.ini:118]: What's the cost that could make it not worth it?  I don't pre-release in most cases but I feel like the world might be better if we all did and we call checked against those.  Maybe.

Comment (graingert) [reply tox.ini:118]: Yeah I think it's worth it, would encourage pre-releases even if none are made now

Comment (nicoddemus) [reply tox.ini:118]: Fair enough, thanks! Done.

**Comment by nicoddemus** [comment]
Anything else? If not it would be nice we could get some approvals here and get this merged. 👍

**Review by hugovk — APPROVED** [review]
This is a great idea! Looks good. I agree on doing light testing of each plugin, this is a quick smoke test* to check nothing exploded. Should catch the major things, and plugins are still encouraged to test against pytest pre-releases.

(*[Etymology](https://stackoverflow.com/a/745279/724176):  'The term originates in hardware repair and has been applied to software. It's intended to be a quick test to see if the application "catches on fire" when run for the first time.')

**Comment by hugovk** [comment]
Oh, one thing to keep in mind: it's possible a new plugin release will break current master, and a new PR will start to fail because of the plugin release, and not the PR changes.

In such cases, it's a good idea to check if a master build also fails in the same way.

It may be needed to temporarily skip smoke testing that plugin to prevent all PRs failing, until either pytest or the plugin has a fix (because it could be a problem on either side, or both!). 

But hopefully this would be rare. And in any case, it's good information to know there's a failure.

File: testing/plugins_integration/pytest_qt_integration.py
Comment (altendky) [inline testing/plugins_integration/pytest_qt_integration.py:1]: Wasn't this supposed to get removed?

Comment (hugovk) [reply testing/plugins_integration/pytest_qt_integration.py:1]: Yep, missed in https://github.com/pytest-dev/pytest/pull/7721/commits/780752b25f976c539f79175b29c43d2ccbc9591e.

**Review by bluetech — APPROVED** [review]
As the person doing most of the breaking, this will be very helpful!

**Review by graingert — APPROVED** [review]

File: testing/plugins_integration/pytest_qt_integration.py
Comment (nicoddemus) [reply testing/plugins_integration/pytest_qt_integration.py:1]: 🤦

Comment (graingert) [reply testing/plugins_integration/pytest_qt_integration.py:1]: Makes me think we should be checking for 100% coverage for the smoke tests

**Comment by nicoddemus** [comment]
Thanks everyone!

**Comment by nicoddemus** [comment]
Of course everyone feel free to add more plugins that they consider relevant. 👍

**Pull request merged**
